src/Workspaces/Core/Portable/Execution/SerializerService_Encoding.cs

```diff
@@ -17,100 +13,103 @@ namespace Microsoft.CodeAnalysis.Serialization
+        private enum EncodingId : byte
+        {
+            None = 0,
+            Named = 1,
+
+            // well-known encodings (parameterized by BOM)
+            UTF8 = 2,
+            UTF8_BOM = 3,
+            UTF32_BE = 4,
+            UTF32_BE_BOM = 5,
+            UTF32_LE = 6,
+            UTF32_LE_BOM = 7,
+            Unicode_BE = 8,
+            Unicode_BE_BOM = 9,
+            Unicode_LE = 10,
+            Unicode_LE_BOM = 11,
+
+            Count
+        }
 
-        private const byte NoEncodingSerialization = 0;
-        private const byte EncodingSerialization = 1;
+        private static readonly Encoding?[] _cachedEncodings = new Encoding[(int)EncodingId.Count];
 
         public static void WriteTo(Encoding? encoding, ObjectWriter writer, CancellationToken cancellationToken)
         {
-            if (encoding == null)
-            {
-                WriteNoEncodingTo(encoding, writer, cancellationToken);
-                return;
-            }
-
             cancellationToken.ThrowIfCancellationRequested();
 
-            var value = GetEncodingBytes(encoding);
-            if (value == null)
+            var kind = GetEncodingKind(encoding);
+            writer.WriteByte((byte)kind);
+
+            if (kind == EncodingId.Named)
             {
-                // we couldn't serialize encoding, act like there is no encoding.
-                WriteNoEncodingTo(encoding, writer, cancellationToken);
-                return;
+                writer.WriteString(encoding!.WebName);
             }
-
-            // write data out
-            writer.WriteByte(EncodingSerialization);
-            writer.WriteValue(value.AsSpan());
         }
 
-        private static void WriteNoEncodingTo(Encoding? encoding, ObjectWriter writer, CancellationToken cancellationToken)
+        private static EncodingId GetEncodingKind(Encoding? encoding)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            writer.WriteByte(NoEncodingSerialization);
-            writer.WriteString(encoding?.WebName);
-        }
-
-        private static byte[]? GetEncodingBytes(Encoding encoding)
-        {
-            try
+            if (encoding == null)
             {
-                if (!s_encodingCache.TryGetValue(encoding, out var value))
-                {
-                    // we don't have cache, cache it
-                    var formatter = new BinaryFormatter();
-                    using var stream = SerializableBytes.CreateWritableStream();
+                return EncodingId.None;
+            }
 
-                    // unfortunately, this is only way to properly clone encoding
-                    formatter.Serialize(stream, encoding);
-                    value = stream.ToArray();
+            switch (encoding.CodePage)
+            {
+                case 1200:
+                    Debug.Assert(HasPreamble(Encoding.Unicode));
```

Comment:
Contract feels more appropraite here.  woudl want to die die die otherwise.

Reply:
The goal here is to avoid allocation of the preamble array. Contract would defeat that.